### PR TITLE
mgr/cephadm: make SMB and NVMEoF upgrade last in staggered upgrade

### PR DESCRIPTION
**Pull request by adk3798**
This needs to happen as some work on the NVMEoF side (still unmerged as of writing this) will make the NVMEoF daemon dependent on the mon. Prior to this patch, in a staggered upgrade, all daemons not using the ceph image were upgraded after the mgr since we typically only care about the default image changing or potential changes to how we handle our systemd units which only needs the mgr to be upgraded to be applied. This NVMEoF dependency on the mon changes this and we can no longer upgrade it directly after the mgr. This patch changes it so the NVMEoF daemon is instead upgraded after all ceph image daemons have been upgraded in a staggered upgrade scenario. Non-staggered upgrades are unaffected as the NVMEoF daemon was already upgraded near the end in that scenario. The SMB dameon has no reason it needs to be upgraded later, but it's in the (small) pool of daemons that don't use the ceph image and aren't for monitoring, so it's been affected by this as well.

NOTE: This is a bit of an ugly patch imo and shows that a refactoring of the upgrade code is likely required. Hopefully this patch is more of a stopgap until that larger effort can be made

Fixes: https://tracker.ceph.com/issues/65809

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/upgrade.py [hunk 1]
```diff
@@ -1214,6 +1216,20 @@ def _do_upgrade(self):
...
+                # which falls in this category, relies on the mons being upgraded as well. This block
+                # sets these daemon types to be upgraded only when all ceph image daemons have been upgraded
+                if any(d in target_digests for d in self.mgr.get_active_mgr_digests()):
+                    ceph_daemons = [d for d in self.mgr.cache.get_daemons() if d.daemon_type in CEPH_IMAGE_TYPES]
+                    _, n1, n2, __ = self._detect_need_upgrade(ceph_daemons, target_digests, target_image)
+                    if not n1 and not n2:
+                        # no ceph daemons need upgrade
+                        dds = [d for d in self.mgr.cache.get_daemons_by_type(
+                            daemon_type) if d.name() not in need_upgrade_names]
+                        _, __, n2, ___ = self._detect_need_upgrade(dds, target_digests, target_image)
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/upgrade.py:1231]: I would have gone for `n2 = self._detect_need_upgrade(...)[2]` but this is fine for now.

**Review by phlogistonjohn — APPROVED** [review]
Looks OK to me.

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-05-09_03:09:40-orch:cephadm-wip-adk-testing-2024-05-08-1927-distro-default-smithi/

failures:
- 21 `in cluster log` type failures that are still being cleaned up, known issue
- 3 mds_upgrade_sequence failures, known issue
- 1 test failed installing `ceph-fuse` ` Status code: 503 for https://mirrors.centos.org/metalink?repo=centos-baseos-9-stream&arch=x86_64&protocol=https,http`, happens on occasion, nothing to block merging over

Overall, nothing unexpected in the run

**Comment by adk3798** [comment]
jenkins test api

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-06-10_15:49:42-orch:cephadm-wip-adk-testing-2024-06-10-0759-distro-default-smithi/

failures consisted of

- 18 `in cluster log` type failures, still being cleaned up and a known issue
- 3 failures in the cleanup of rados_api_tests, don't know the tracker number but this is a known issue
- 2 failures in mds_upgrade_sequence, known issue
- 1 failure with staggered upgrade while agent is active, known issue
- lots of dead jobs due to `Error reimaging machines: Failed to power on smithiXXX`, general lab/teuthology issue unrelated to the PRs in the run

Nothing to block merging for most PRs in the run

**Pull request merged**